crates/bevy_tasks/src/executor.rs

```diff
@@ -135,34 +135,46 @@ impl<'a> LocalExecutor<'a> {
     /// Creates a single-threaded executor.
     pub fn new() -> LocalExecutor<'a> {
         LocalExecutor {
-            inner: Executor::new(&[1]),
+            queue: ConcurrentQueue::unbounded(),
             _marker: PhantomData,
         }
     }
 
     /// Spawns a task onto the executor.
     pub fn spawn<T: 'a>(&self, future: impl Future<Output = T> + 'a) -> Task<T> {
-        // Create the task and schedule it
+        // SAFETY: The spawned Task can only be progressed via `try_tick` which must be accessed
+        // from the thread that owns the executor and the task.
+        //
+        // Even if the returned Task and waker are sent to another thread, the associated inner
+        // task is only dropped when `try_tick` is triggered.
         let (runnable, task) = unsafe { async_task::spawn_unchecked(future, self.schedule()) };
-        runnable.schedule();
+        // SAFETY: The queue is unbounded, this can never fail.
+        unsafe {
+            self.queue.push(runnable).unwrap_unchecked();
+        }
         task
     }
 
     /// Attempts to run a task if at least one is scheduled.
     ///
     /// Running a scheduled task means simply polling its future once.
     pub fn try_tick(&self) -> bool {
-        self.inner.try_tick(0)
+        match self.queue.pop() {
+            Err(_) => false,
+            Ok(runnable) => {
+                runnable.run();
+                true
+            }
+        }
     }
 
     /// Returns a function that schedules a runnable task when it gets woken up.
-    fn schedule(&self) -> impl Fn(Runnable) + Send + Sync + 'static {
-        let state = self.inner.state.clone();
-
+    fn schedule(&self) -> impl Fn(Runnable) + '_ {
         move |runnable| {
-            let group = &state.groups[0];
-            group.queue.push(runnable).unwrap();
-            group.notify();
+            // SAFETY: The queue is unbounded, this can never fail.
```

Comment:
Do we need to be concerned at all about integer overflow here?

Reply:
The unbounded queue is implemented as a atomic linked list of blocks of 32 elements, so it actually never can overflow.

Reply:
Is there any benefit to avoiding the unwrap here?

If there's an appreciable gain, I'd be fine with just expanding this safety comment some more.

Reply:
This does happen in a "tight" loop in `TaskPool::scope` depending on how long systems take, and while spawning tasks. However, I think those are largely dominated by the atomic accesses in the queue.

I'd love to replace it with a `RefCell<VecDeque<Runnable>>`, but that prevents users from spawning tasks from inside a local executor tick.